File: cli/internal_testsuites/typeScript/src/testsuite/network_partition_test/network_partition.test.ts
import { AddPersonArgs, IncrementBooksReadArgs } from "example-api-server-api-lib";
import { EnclaveContext, PartitionID, ServiceID, BlockedPartitionConnection, UnblockedPartitionConnection, PartitionConnections } from "kurtosis-core-api-lib";
import log from "loglevel"
import { Result, ok, err } from "neverthrow"
import * as grpc from "@grpc/grpc-js"
import * as google_protobuf_empty_pb from "google-protobuf/google/protobuf/empty_pb";

import { createEnclave } from "../../test_helpers/enclave_setup";
import { addAPIService, addAPIServiceToPartition, addDatastoreService } from "../../test_helpers/test_helpers";
import { PartitionConnection } from "kurtosis-core-api-lib/build/lib/enclaves/partition_connection";

const TEST_NAME = "network-partition-test";
const IS_PARTITIONING_ENABLED = true;

const API_PARTITION_ID: PartitionID = "api";
const DATASTORE_PARTITION_ID: PartitionID = "datastore";

const DATASTORE_SERVICE_ID: ServiceID   = "datastore";
const API1_SERVICE_ID: ServiceID   = "api1";
const API2_SERVICE_ID: ServiceID   = "api2";

const TEST_PERSON_ID = "46";

const MILLISECONDS_IN_A_SECOND = 1000;
const CONTEXT_TIME_OUT = 2 * MILLISECONDS_IN_A_SECOND;

jest.setTimeout(30000)

test("Test network partition", async () => {
     // ------------------------------------- ENGINE SETUP ----------------------------------------------
     const createEnclaveResult = await createEnclave(TEST_NAME, IS_PARTITIONING_ENABLED)

     if(createEnclaveResult.isErr()) { throw createEnclaveResult.error }
 
     const { enclaveContext, stopEnclaveFunction } = createEnclaveResult.value
 
     try {

        // ------------------------------------- TEST SETUP ----------------------------------------------

        const addDatastoreServiceResult = await addDatastoreService(DATASTORE_SERVICE_ID, enclaveContext)

        if(addDatastoreServiceResult.isErr()) { throw addDatastoreServiceResult.error }

        const { 
            serviceContext: datastoreServiceContext, 
            clientCloseFunction:datastoreClientCloseFunction 
        } = addDatastoreServiceResult.value

        try {

            const addAPIServiceResult = await addAPIService(API1_SERVICE_ID, enclaveContext, datastoreServiceContext.getPrivateIPAddress())
            if(addAPIServiceResult.isErr()){ throw addAPIServiceResult.error }

            const { client: apiClient1, clientCloseFunction: apiClient1CloseFunction } = addAPIServiceResult.value;

            const addPersonArgs = new AddPersonArgs()
            addPersonArgs.setPersonId(TEST_PERSON_ID)

            const addPerson1ResultPromise: Promise<Result<google_protobuf_empty_pb.Empty, Error>> = new Promise((resolve, _unusedReject) => {
                apiClient1.addPerson(addPersonArgs, (error: grpc.ServiceError | null, response?: google_protobuf_empty_pb.Empty) => {
                    if (error === null) {
                        if (!response) {
                            resolve(err(new Error("No error was encountered but the response was still falsy; this should never happen")));
                        } else {
                            resolve(ok(response!));
                        }
                    } else {
                        resolve(err(error));
                    }
                })
            })

            try {
                const addPerson1Result = await addPerson1ResultPromise;
                if(addPerson1Result.isErr()) {
                    log.error(`An error occurred adding test person with ID ${TEST_PERSON_ID}`)
                    throw addPerson1Result.error;
                }

                const incrementBooksReadArgs = new IncrementBooksReadArgs()
                incrementBooksReadArgs.setPersonId(TEST_PERSON_ID)
            
                const incrementBooksReadResultPromise: Promise<Result<google_protobuf_empty_pb.Empty, Error>> = new Promise((resolve, _unusedReject) => {
                    apiClient1.incrementBooksRead(incrementBooksReadArgs, (error: grpc.ServiceError | null, response?: google_protobuf_empty_pb.Empty) => {
                        if (error === null) {
                            if (!response) {
                                resolve(err(new Error("No error was encountered but the response was still falsy; this should never happen")));
                            } else {
                                resolve(ok(response!));
                            }
                        } else {
                            resolve(err(error));
                        }
                    })
                })

                const incrementBooksReadResult = await incrementBooksReadResultPromise;
                if(incrementBooksReadResult.isErr()) { 
                    log.error("An error occurred test person's books read in preparation for the test")
                    throw incrementBooksReadResult.error;
                }

                // ------------------------------------- TEST RUN ----------------------------------------------
                
                log.info("Partitioning API and datastore services off from each other...")
                const repartitionNetworkResult =  await repartitionNetwork(enclaveContext, true, false)
                if(repartitionNetworkResult.isErr()){
                    log.error("An error occurred repartitioning the network to block access between API <-> datastore")
                    throw repartitionNetworkResult.error
                }
                
                log.info("Repartition complete")

                log.info("Incrementing books read via API 1 while partition is in place, to verify no comms are possible...")

                const newIncrementBooksReadResult = await incrementBooksReadResultPromise;
                if(newIncrementBooksReadResult.isOk()){
                    log.error("Expected the book increment call via API 1 to fail due to the network " +
                    "partition between API and datastore services, but no error was thrown")
                    return err(newIncrementBooksReadResult.value)
                }
                    
                log.info(`Incrementing books read via API 1 threw the following error as expected due to network partition: ${newIncrementBooksReadResult.error}`)
                    
                // Adding another API service while the partition is in place ensures that partitioning works even when you add a node
	            log.info("Adding second API container, to ensure adding a service under partition works...")

                const addAPIServiceToPartitionResult = await addAPIServiceToPartition(API2_SERVICE_ID, enclaveContext, datastoreServiceContext.getPrivateIPAddress(), API_PARTITION_ID)

                if(addAPIServiceToPartitionResult.isErr()){
                    log.error("An error occurred adding the second API service to the network")
                    throw addAPIServiceToPartitionResult.error
                }

                const { client: apiClient2, clientCloseFunction: apiClient2CloseFunction } = addAPIServiceToPartitionResult.value

            	log.info("Second API container added successfully")

                log.info("Incrementing books read via API 2 while partition is in place, to verify no comms are possible...")

                try {

                    const incrementBooksRead2ResultPromise: Promise<Result<google_protobuf_empty_pb.Empty, Error>> = new Promise((resolve, _unusedReject) => {
                        apiClient2.incrementBooksRead(incrementBooksReadArgs, (error: grpc.ServiceError | null, response?: google_protobuf_empty_pb.Empty) => {
                            if (error === null) {
                                if (!response) {
                                    resolve(err(new Error("No error was encountered but the response was still falsy; this should never happen")));
                                } else {
                                    resolve(ok(response!));
                                }
                            } else {
                                resolve(err(error));
                            }
                        })
                    })
    
                    const incrementBooksRead2Result = await incrementBooksRead2ResultPromise;
                    if(incrementBooksRead2Result.isOk()) { 
                        log.error("Expected the book increment call via API 2 to fail due to the network " +
                        "partition between API and datastore services, but no error was thrown")
                        throw incrementBooksRead2Result.value
                    }

                    log.info(`Incrementing books read via API 2 threw the following error as expected due to network partition: ${incrementBooksRead2Result.error}`)


                    // Now, open the network back up
                    log.info("Repartitioning to heal partition between API and datastore...")

                    const repartition2NetworkResult = await repartitionNetwork(enclaveContext, false, true)
                    if(repartition2NetworkResult.isErr()){
                        log.error("An error occurred healing the partition")
                        throw repartition2NetworkResult.error
                    }

                    log.info("Partition healed successfully")

                    log.info("Making another call via API 1 to increment books read, to ensure the partition is open...")
                    // Use infinite timeout because we expect the partition healing to fix the issue
                    



                }finally{
                    apiClient2CloseFunction()
                }
            }finally{
                apiClient1CloseFunction()
            }
        }finally{
            datastoreClientCloseFunction()
        }
     }finally{
         stopEnclaveFunction()
     }

     jest.clearAllTimers()
})

/*
Creates a repartitioner that will partition the network between the API & datastore services, with the connection between them configurable
*/

async function repartitionNetwork( enclaveContext: EnclaveContext,  isConnectionBlocked: boolean, isApi2ServiceAddedYet: boolean ): Promise<Result<null, Error>> {
    
    const apiPartitionServiceIds = new Set<ServiceID>()
    apiPartitionServiceIds.add(API1_SERVICE_ID)

    if(isApi2ServiceAddedYet){
        apiPartitionServiceIds.add(API2_SERVICE_ID)
    }

    const connectionBetweenPartitions: PartitionConnection = isConnectionBlocked ? new BlockedPartitionConnection() : new UnblockedPartitionConnection()
    
    
    const partitionServices = new Map<PartitionID, Set<ServiceID>>()
    const datastoreServiceId = new Set<ServiceID>(DATASTORE_SERVICE_ID)

    partitionServices.set(API_PARTITION_ID, apiPartitionServiceIds)
    partitionServices.set(DATASTORE_PARTITION_ID, datastoreServiceId)

    const partitionConnections = new Map<PartitionID, Map<PartitionID, PartitionConnection>>()
    const datastorePartitionId = new Map<PartitionID, PartitionConnection>([[DATASTORE_PARTITION_ID,connectionBetweenPartitions]]);
    partitionConnections.set(API_PARTITION_ID, datastorePartitionId);

    const defaultPartitionConnection = new UnblockedPartitionConnection()

    const repartitionNetworkResult = await enclaveContext.repartitionNetwork(partitionServices, partitionConnections, defaultPartitionConnection);

    if(repartitionNetworkResult.isErr()){
        log.error(`An error occurred repartitioning the network with isConnectionBlocked = ${isConnectionBlocked}`)
        return err(repartitionNetworkResult.error)
    }

    return ok(null)
}

File: cli/internal_testsuites/golang/testsuite/network_partition_test/network_partition_test.go
//+build !minikube

// We don't run this test in Kubernetes because, as of 2022-07-07, Kubernetes doesn't support network partitioning

package network_partition_test

import (
	"context"
	"github.com/kurtosis-tech/example-api-server/api/golang/example_api_server_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-cli/golang_internal_testsuite/test_helpers"
	"github.com/kurtosis-tech/kurtosis-core-api-lib/api/golang/lib/enclaves"
	"github.com/kurtosis-tech/kurtosis-core-api-lib/api/golang/lib/services"
	"github.com/kurtosis-tech/stacktrace"
	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/require"
	"testing"
	"time"
)

const (
	testName              = "network-partition"
	isPartitioningEnabled = true

	apiPartitionId       enclaves.PartitionID = "api"
	datastorePartitionId enclaves.PartitionID = "datastore"

	datastoreServiceId services.ServiceID = "datastore"
	api1ServiceId      services.ServiceID = "api1"
	api2ServiceId      services.ServiceID = "api2"

	testPersonId   = "46"
	contextTimeOut = 2 * time.Second
)

func TestNetworkPartition(t *testing.T) {
	ctx := context.Background()

	// ------------------------------------- ENGINE SETUP ----------------------------------------------
	enclaveCtx, stopEnclaveFunc, _, err := test_helpers.CreateEnclave(t, ctx, testName, isPartitioningEnabled)
	require.NoError(t, err, "An error occurred creating an enclave")
	defer stopEnclaveFunc()

	// ------------------------------------- TEST SETUP ----------------------------------------------

	datastoreServiceCtx, _, datastoreClientCloseFunc, err := test_helpers.AddDatastoreService(ctx, datastoreServiceId, enclaveCtx)
	require.NoError(t, err, "An error occurred adding the datastore service")
	defer datastoreClientCloseFunc()

	_, api1Client, api1ClientCloseFunc, err := test_helpers.AddAPIService(ctx, api1ServiceId, enclaveCtx, datastoreServiceCtx.GetPrivateIPAddress())
	require.NoError(t, err, "An error occurred adding the first API service")
	defer api1ClientCloseFunc()

	addPersonArgs := &example_api_server_rpc_api_bindings.AddPersonArgs{
		PersonId: testPersonId,
	}
	_, err = api1Client.AddPerson(ctx, addPersonArgs)
	require.NoError(t, err, "An error occurred adding test person with ID '%v'", testPersonId)

	incrementBooksReadArgs := &example_api_server_rpc_api_bindings.IncrementBooksReadArgs{
		PersonId: testPersonId,
	}
	_, err = api1Client.IncrementBooksRead(ctx, incrementBooksReadArgs)
	require.NoError(t, err, "An error occurred test person's books read in preparation for the test")

	// ------------------------------------- TEST RUN ----------------------------------------------
	logrus.Info("Partitioning API and datastore services off from each other...")
	require.NoError(t, repartitionNetwork(enclaveCtx, true, false), "An error occurred repartitioning the network to block access between API <-> datastore")
	logrus.Info("Repartition complete")

	logrus.Info("Incrementing books read via API 1 while partition is in place, to verify no comms are possible...")
	ctxWithTimeOut, cancelCtxWithTimeOut := context.WithTimeout(ctx, contextTimeOut)
	defer cancelCtxWithTimeOut()
	_, err = api1Client.IncrementBooksRead(ctxWithTimeOut, incrementBooksReadArgs)
	require.Error(t, err, "Expected the book increment call via API 1 to fail due to the network "+
		"partition between API and datastore services, but no error was thrown")
	logrus.Infof("Incrementing books read via API 1 threw the following error as expected due to network partition: %v", err)

	// Adding another API service while the partition is in place ensures that partitioning works even when you add a node
	logrus.Info("Adding second API container, to ensure adding a service under partition works...")
	_, apiClient2, apiClient2ConnCloseFunc, err := test_helpers.AddAPIServiceToPartition(
		ctx,
		api2ServiceId,
		enclaveCtx,
		datastoreServiceCtx.GetPrivateIPAddress(),
		apiPartitionId,
	)
	require.NoError(t, err, "An error occurred adding the second API service to the network")
	defer apiClient2ConnCloseFunc()
	logrus.Info("Second API container added successfully")

	logrus.Info("Incrementing books read via API 2 while partition is in place, to verify no comms are possible...")
	ctxWithTimeOut2, cancelCtxWithTimeOut2 := context.WithTimeout(ctx, contextTimeOut)
	defer cancelCtxWithTimeOut2()
	_, err = apiClient2.IncrementBooksRead(ctxWithTimeOut2, incrementBooksReadArgs)
	require.Error(t, err, "Expected the book increment call via API 2 to fail due to the network "+
		"partition between API and datastore services, but no error was thrown")
	logrus.Infof("Incrementing books read via API 2 threw the following error as expected due to network partition: %v", err)

	// Now, open the network back up
	logrus.Info("Repartitioning to heal partition between API and datastore...")
	require.NoError(t, repartitionNetwork(enclaveCtx, false, true), "An error occurred healing the partition")
	logrus.Info("Partition healed successfully")

	logrus.Info("Making another call via API 1 to increment books read, to ensure the partition is open...")
	// Use infinite timeout because we expect the partition healing to fix the issue
	_, err = api1Client.IncrementBooksRead(ctx, incrementBooksReadArgs)
	require.NoError(t, err, "An error occurred incrementing the number of books read via API 1, even though the partition should have been "+
		"healed by the goroutine")
	logrus.Info("Successfully incremented books read via API 1, indicating that the partition has healed successfully!")

	logrus.Info("Making another call via API 2 to increment books read, to ensure the partition is open...")
	// Use infinite timeout because we expect the partition healing to fix the issue
	_, err = apiClient2.IncrementBooksRead(ctx, incrementBooksReadArgs)
	require.NoError(t, err, "An error occurred incrementing the number of books read via API 2, even though the partition should have been "+
		"healed by the goroutine")
	logrus.Info("Successfully incremented books read via API 2, indicating that the partition has healed successfully!")
}

/*
Creates a repartitioner that will partition the network between the API & datastore services, with the connection between them configurable
*/
func repartitionNetwork(
	enclaveCtx *enclaves.EnclaveContext,
	isConnectionBlocked bool,
	isApi2ServiceAddedYet bool,
) error {
	apiPartitionServiceIds := map[services.ServiceID]bool{
		api1ServiceId: true,
	}
	if isApi2ServiceAddedYet {
		apiPartitionServiceIds[api2ServiceId] = true
	}

	var connectionBetweenPartitions enclaves.PartitionConnection
	if isConnectionBlocked {
		connectionBetweenPartitions = enclaves.NewBlockedPartitionConnection()
	} else {
		connectionBetweenPartitions = enclaves.NewUnblockedPartitionConnection()
	}
	partitionServices := map[enclaves.PartitionID]map[services.ServiceID]bool{
		apiPartitionId: apiPartitionServiceIds,
		datastorePartitionId: {
			datastoreServiceId: true,
		},
	}
	partitionConnections := map[enclaves.PartitionID]map[enclaves.PartitionID]enclaves.PartitionConnection{
		apiPartitionId: {
			datastorePartitionId: connectionBetweenPartitions,
		},
	}
	defaultPartitionConnection := enclaves.NewUnblockedPartitionConnection()
	if err := enclaveCtx.RepartitionNetwork(partitionServices, partitionConnections, defaultPartitionConnection); err != nil {
		return stacktrace.Propagate(
			err,
			"An error occurred repartitioning the network with isConnectionBlocked = %v",
			isConnectionBlocked,
		)
	}
	return nil
}
